mtt/src/mtt.rs

```diff
@@ -0,0 +1,716 @@
+        L2_1G_NC_TYPE << L2_TYPE_SHIFT
+    }
+}
+
+struct Mtt64MEntry<'a> {
+    entry: &'a mut u64,
+}
+
+impl<'a> Mtt64MEntry<'a> {
+    fn new(entry: &'a mut u64) -> Mtt64MEntry {
+        Self { entry }
+    }
+
+    fn raw_conf_64m() -> u64 {
+        L2_64M_TYPE << L2_TYPE_SHIFT | 0xffff_ffff
+    }
+
+    fn raw_non_conf_64m() -> u64 {
+        L2_64M_TYPE << L2_TYPE_SHIFT
+    }
+
+    fn get_bit_index(addr: SupervisorPageAddr) -> u64 {
+        let addr_bits = PageSize::Size2M.round_down(addr.bits());
+        let round_down_64m = addr_bits & !((1024 * 1024 * 64) - 1);
+        (addr_bits - round_down_64m) / PageSize::Size2M as u64
+    }
+
+    /// # Safety
+    /// The caller must ensure that is is safe to destroy the contents of the entire 2MB range
+    /// spanned by `addr`.
+    unsafe fn set_memory_type(&mut self, addr: SupervisorPageAddr, memory_type: MttMemoryType) {
+        let bit_index = Self::get_bit_index(addr);
+        if memory_type == MttMemoryType::Confidential {
+            *self.entry |= 1 << bit_index;
+        } else {
+            *self.entry &= !(1 << bit_index);
+        }
+    }
+
+    fn is_conf(&self, addr: SupervisorPageAddr) -> bool {
+        let bit_index = Self::get_bit_index(addr);
+        *self.entry & (1 << bit_index) != 0
+    }
+
+    // Converts an existing MTT L2 entry representing a 64MB range into a MttL1Entry.
+    // The new entry will span the same 64MB range, but it will be broken up into
+    // 32-regions of 2MB each. Each of the 2MB regions is mapped to 16 8-bytes entries
+    // with 2-bits per 4K page in the region.
+    // Note that it's OK to update the L1 page without any concern about a potential
+    // race condition with an hardware initiated walk of the MTT. This is because we
+    // bare converting from MTT2BPages to MTTL1Dir, and the type of MTT L2 entry remains
+    // unchanged until it's atomically updated below following the L1 page write.
+    fn split_to_l1_entry(self, l1_page: Page<InternalClean>) {
+        let l1_page_addr = l1_page.as_bytes().as_ptr() as *mut u64;
+        // Safety: We are writing in a page that's uniquely owned
+        // and will write no more than 4K bytes.
+        let l1_slice =
+            unsafe { &mut *core::slice::from_raw_parts_mut(l1_page_addr, L1_ENTRY_COUNT) };
+        let value = *self.entry;
+        for i in 0usize..=31 {
+            // Mark the 2MB sub-range as confidential if the bit in the 32-bit mask is set
+            let write_value = if (value & (1 << i)) != 0 {
+                // This represents 32 4K-pages of confidential memory in a L1-page
+                0x5555_5555_5555_5555u64
+            } else {
+                0
+            };
+            l1_slice.iter_mut().skip(i * 16).take(16).for_each(|entry| {
+                *entry = write_value;
+            });
+        }
+        let pfn = SupervisorPageAddr::new(RawAddr::supervisor(l1_page_addr as u64))
+            .unwrap()
+            .pfn();
```

Comment:
No idea why github keeps losing my comment, but, but `l1_page.pfn()` (actually it's `l1_page.addr.pfn()` can't be accessed as it's private. We can fix this in a follow-up PR since cramming a change for that seems odd.

Reply:
huh? `pfn()` is a method of the `PhysPage` trait, which `Page` implements, and thus should be visible. and even if it weren't, `pfn()` is a public method of `PageAddr<>`. what am i missing?

Reply:
Turns out that I was misled by a Rust Analyzer gotcha. It's usually very reliable, but in this particular case, it was indicating that pfn() wasn't accessible, until after I explicitly coaxed it into looking up the PhysPage trait. The lesson is to trust your tools, but not too much :).